aspnetcore/mvc/razor-pages/razor-pages-convention-features.md

```diff
@@ -0,0 +1,290 @@
+---
+title: Razor Pages route and app convention features in ASP.NET Core
+author: guardrex
+description: Discover how route and app model provider convention features help you control page routing, discovery, and processing.
+keywords: ASP.NET Core,Razor Pages,conventions,AddFolderRouteModelConvention,AddPageRouteModelConvention,AddPageRoute,AddFolderApplicationModelConvention,AddPageApplicationModelConvention,ConfigureFilter,filters
+ms.author: riande
+manager: wpickett
+ms.date: 10/23/2017
+ms.topic: article
+ms.assetid: 6b60514c-81ad-485b-bb22-9b71416dff08
+ms.technology: aspnet
+ms.prod: asp.net-core
+uid: mvc/razor-pages/razor-pages-convention-features
+---
+# Razor Pages route and app convention features in ASP.NET Core
+
+By [Luke Latham](https://github.com/guardrex)
+
+Learn how to use page route and app model provider convention features to control page routing, discovery, and processing in Razor Pages apps. When you need to configure custom page routes for individual pages, configure routing to pages with the [AddPageRoute convention](#configure-a-page-route) described later in this topic.
+
+Use the [sample app](https://github.com/aspnet/Docs/tree/master/aspnetcore/mvc/razor-pages/razor-pages-convention-features/sample) ([how to download](xref:tutorials/index#how-to-download-a-sample)) to explore the features described in this topic.
+
+| Features | The sample demonstrates ... |
+| -------- | --------------------------- |
+| [Route and app model conventions](#add-route-and-app-model-conventions) | Adding a route attribute and header to an app's pages. |
+| [Page route action conventions](#page-route-action-conventions) | Adding a route attribute to pages in a folder and to a single page. |
+| [Page model action conventions](#page-model-action-conventions) | Adding a header to pages in a folder, adding a header to a single page, and configuring a filter factory to add a header to an app's pages. |
+| [Default page app model provider](#replace-the-default-page-app-model-provider) | Replacing the default page model provider to change the conventions for handler naming. |
+
+## Add route and app model conventions
+
+Add a delegate for [IPageConvention](/dotnet/api/microsoft.aspnetcore.mvc.applicationmodels.ipageconvention) to add route and app model conventions that apply to Razor Pages.
+
+**Add a route model convention to all pages**
+
+Use [Conventions](/dotnet/api/microsoft.aspnetcore.mvc.razorpages.razorpagesoptions.conventions) to create and add an [IPageRouteModelConvention](/dotnet/api/microsoft.aspnetcore.mvc.applicationmodels.ipageroutemodelconvention) to the collection of [IPageConvention](/dotnet/api/microsoft.aspnetcore.mvc.applicationmodels.ipageconvention) instances that are applied during route and page model construction.
+
+The sample app adds a `{globalAttribute?}` optional route attribute to all of the pages in the app:
```

Comment:
@pranavkm Is "route attribute" the correct term to describe `{globalAttribute?}`? Or is this better described as a "route constraint"?

Reply:
I'm glad u asked ... `AttributeRouteModel` is the model for the `[Route()]` attribute, and it really looks like these (e.g., `{globalAttribute?}`) should be called "route templates." *One case of fine Belgian ale for @pranavkm if he'll tell us the answer!* :beer: lol

Reply:
Route template would be the correct term. You might need to update the type too (`GobalAttributePageRouteModelConvention -> GlobalTemplate....)